File: sdks/local-rp/ruby/test/test_begin.rb
# frozen_string_literal: true

require_relative 'test_helper'

class TestBegin < Minitest::Test
  def setup
    @now = Time.utc(2026, 1, 1)
    @material = LinkkeysLocalRp.generate_local_rp_identity(
      LinkkeysLocalRp::Identity::GenerateLocalRpIdentityConfig.new(app_name: 'Test App', now: @now)
    )
  end

  def begin_login(identity)
    LinkkeysLocalRp.begin_local_login(LinkkeysLocalRp::Begin::BeginLocalLoginConfig.new(
      key_material: @material, callback_url: 'http://localhost/callback', user_domain: identity, now: @now
    ))
  end

  def test_identity_input_controls_destination_and_username_hint
    redirect, pending = begin_login('Alice+work@ID.Example.TEST')
    assert redirect.redirect_url.end_with?('&username=Alice%2Bwork')
    assert_equal 'id.example.test', pending.user_domain
    redirect, = begin_login('example.test')
    refute_includes redirect.redirect_url, 'username='
  end

  def test_malformed_identity_input_is_rejected
    %w[alice alice@@example.test https://example.test].each do |identity|
      assert_raises(LinkkeysLocalRp::Begin::Error) { begin_login(identity) }
    end
  end
end

File: sdks/local-rp/python/tests/test_begin.py
from datetime import datetime, timezone
from urllib.parse import parse_qs, urlparse

import pytest

from linkkeys_local_rp.begin import BeginLocalLoginConfig, BeginLoginError, begin_local_login
from linkkeys_local_rp.identity import GenerateLocalRpIdentityConfig, generate_local_rp_identity


NOW = datetime(2026, 1, 1, tzinfo=timezone.utc)
KEY_MATERIAL = generate_local_rp_identity(GenerateLocalRpIdentityConfig(app_name="Test App", now=NOW))


def begin(identity: str):
    return begin_local_login(BeginLocalLoginConfig(KEY_MATERIAL, "http://localhost/callback", identity, NOW))


def test_full_login_adds_username_hint_and_bare_domain_does_not():
    redirect, pending = begin("Alice+work@ID.Example.TEST")
    assert parse_qs(urlparse(redirect.redirect_url).query)["username"] == ["Alice+work"]
    assert pending.user_domain == "id.example.test"
    redirect, _ = begin("example.test")
    assert "username" not in parse_qs(urlparse(redirect.redirect_url).query)


@pytest.mark.parametrize("identity", ["alice", "alice@@example.test", "https://example.test"])
def test_malformed_identity_is_rejected(identity):
    with pytest.raises(BeginLoginError):
        begin(identity)
